### ISSUE TITLE
Lighten the folder structure

### ISSUE THREAD

**Issue by albestro** [eth-cscs/DLA-Future#236]
@teonnik pointed out during his reviews that the tree structure of algorithms (i.e. factorization, solver, auxiliary) can be lighten a bit.

Mainly the proposals are about:
- CMakeFiles organization for `test/unit` (https://github.com/eth-cscs/DLA-Future/pull/224#discussion_r427290606 and https://github.com/eth-cscs/DLA-Future/pull/224#discussion_r427291333)
- Evaluate moving implementation details to `src` (https://github.com/eth-cscs/DLA-Future/pull/224#discussion_r427273595)

As soon as we decide how to streamline, we can proceed on changing all existing instances at the same way in one shot.

**Comment by teonnik** [comment]
Do we have consensus on the following suggestions or do we want some changes? :

- Move contents of `.tpp` files into corresponding `.cpp` files before the ETI. If there are multiple `.tpp` corresponding to a `.h`, move the contents of all `.tpp` to the `.cpp` (e.g. `matrix.h` with `matrix.tpp`, `matrix_const.tpp` to `matrix.cpp`). 
- Headers only used within `.cpp` files are private/internal and should be in the `src` directory.
- Headers that are private/internal but are used within our public headers should stay in the `include/dlaf` directory.

The current structure was discussed here : https://github.com/eth-cscs/DLA-Future/pull/146 

There are a few issues I have with it:

- deeply nested hierarchies with a lot of small, "glue" files like `factorization/internal.h`. 
- often several directory layers have to be navigated with similarly named files/folders (e.g. `mc`/`mc.h`/`mc.tpp`) to get to the actual files containing the implementation of the algorithms. 
- The structure is designed to have separate `mc` (multicore) and e.g. `gpu` (or other backends) hierarchies, however it is not clear yet whether the separation is beneficial as it is not clear whether the `mc`/`gpu` back ends will have divergent or similar implementations.
- overuse of namespaces : it's good to group things into namespaces but we should use them judiciously and for their main purpose : to avoid name clashes https://stackoverflow.com/questions/713698/c-namespaces-advice
- using struct with static methods when function overloading is sufficient (e.g. in `factorization/mc.h`)

For example, instead of this:

```
solver
├── internal.h
├── mc.h
└── triangular/
    ├── mc
    │   ├── triangular_LLN.h
    │   ├── triangular_LLT.h
    │   ├── triangular_LUN.h
    │   ├── triangular_LUT.h
    │   ├── triangular_RLN.h
    │   ├── triangular_RLT.h
    │   ├── triangular_RUN.h
    │   └── triangular_RUT.h
    └── mc.tpp
```

why not this:

```
solver/
├── triangular.h
└── triangular/
    ├── LLN.h
    ├── LLT.h
    ├── LUN.h
    ├── LUT.h
    ├── RLN.h
    ├── RLT.h
    ├── RUN.h
    └── RUT.h
```

or instead of this:

```
factorization/
├── cholesky/
│   ├── mc/
│   │   └── cholesky_L.h
│   └── mc.tpp
├── internal.h
└── mc.h
```

this: 

```
factorization/
├── cholesky.h 
├── cholesky/
│   └── lower.h
```

**Comment by teonnik** [comment]
> Move contents of .tpp files into corresponding .cpp files before the ETI. If there are multiple .tpp corresponding to a .h, move the contents of all .tpp to the .cpp (e.g. matrix.h with matrix.tpp, matrix_const.tpp to matrix.cpp)

Regarding this item, I'd actually be fine if we leave things as they are or moving the contents of the `.tpp` files within the `.h` files (separation of implementation and declaration can still be maintained there). This will allow us to potentially support truly generic Scalar types (not just `float`, `double`, `std::complex<float>` and `std::complex<double>`) in the future. Moreover, if we decide to split the ETI for each Scalar datatype into separate `.cpp` (e.g. tile_z.cpp / tile_d.cpp, ... etc) files in the future (for e.g. improved incremental compile times), we won't have to deal with the question of where do we put the contents of the former `.tpp` file from the now split `.cpp` file.

**Comment by rasolca** [comment]
> Do we have consensus on the following suggestions or do we want some changes? :
> 
>     * Move contents of `.tpp` files into corresponding `.cpp` files before the ETI. If there are multiple `.tpp` corresponding to a `.h`, move the contents of all `.tpp` to the `.cpp` (e.g. `matrix.h` with `matrix.tpp`, `matrix_const.tpp` to `matrix.cpp`).

This is not possible all the times, since a template member function of a template class may exist.
(I'm facing this problem with matrix_views).
Moreover I think that using ETI for matrices and tiles doesn't improve compile time, as only a few members are not inlined.

> 
>     * Headers only used within `.cpp` files are private/internal and should be in the `src` directory.
>     * Headers that are private/internal but are used within our public headers should stay in the `include/dlaf` directory.
> 
> 
> The current structure (for solver/factorization) was discussed here : #146
> 
> There are a few issues I have with it:
> 
>     * deeply nested hierarchies with a lot of small, "glue" files like `factorization/internal.h`.
The general template should be defined somewhere.

> 
>     * often several directory layers have to be navigated with similarly named files/folders (e.g. `mc`/`mc.h`/`mc.tpp`) to get to the actual files containing the implementation of the algorithms.
> 
>     * The structure is designed to have separate `mc` (multicore) and e.g. `gpu` (or other backends) hierarchies, however it is not clear yet whether the separation is beneficial as it is not clear whether the `mc`/`gpu` back ends will have divergent or similar implementations. 

I'm sure that we need specific GPU implementations.

>     * overuse of namespaces : it's good to group things into namespaces but we should use them judiciously and for their main purpose : to avoid name clashes https://stackoverflow.com/questions/713698/c-namespaces-advice

Currently namespaces are limited to dlaf::<category>::<internal/test/util>. Did I miss overuse somewhere?

>     * using struct with static methods when function overloading is sufficient (e.g. in `factorization/mc.h`)

Overload will not be sufficient. It should be possible to use the GPU backend even when the initial data is on CPU memory. Moreover two devices may share the same memory (e.g. APUs)

The idea is to provide 2 level of functions:
- level1: the user specifies no backend, which will be chosen at run time depending on some parameters e.g. cpu/gpu memory, size of the problem ,... (not yet available)
- level2: the user specifies which backend to use (through the struct), e.g. data is on CPU but computation should be done on GPU.

> For example, instead of this:
> 
> ```
> solver
> ├── internal.h
> ├── mc.h
> └── triangular/
>     ├── mc
>     │   ├── triangular_LLN.h
>     │   ├── triangular_LLT.h
>     │   ├── triangular_LUN.h
>     │   ├── triangular_LUT.h
>     │   ├── triangular_RLN.h
>     │   ├── triangular_RLT.h
>     │   ├── triangular_RUN.h
>     │   └── triangular_RUT.h
>     └── mc.tpp
> ```
> 
> why not this:
> 
> ```
> solver/
> ├── triangular.h
> └── triangular/
>     ├── LLN.h
>     ├── LLT.h
>     ├── LUN.h
>     ├── LUT.h
>     ├── RLN.h
>     ├── RLT.h
>     ├── RUN.h
>     └── RUT.h
> ```
> 
> or instead of this:
> 
> ```
> factorization/
> ├── cholesky/
> │   ├── mc/
> │   │   └── cholesky_L.h
> │   └── mc.tpp
> ├── internal.h
> └── mc.h
> ```
> 
> this:
> 
> ```
> factorization/
> ├── cholesky.h 
> ├── cholesky/
> │   └── lower.h
> ```

the idea is to have the following structure (currently is some parts are missing):
(Example with factorizations: (LU and Cholesky))
```
factorization/
  - factorization.h # includes the backends and declarations of level1 functions.
  - mc.h # declaration of mc backend functions (level2)
  - gpu.h # declaration of gpu backend functions (level2)
  - cholesky/
     - cholesky.tpp # implementation of level1 functions (i.e. the logic that chose which level2 has to be called)
     - mc.tpp # implementation of level2 functions (mc backend) (dispatcher for internal functions)
     - mc/*.h # implementation of internal functions of mc backend
     - gpu/*.h # implementation of internal functions of gpu backend
  - lu/
     - lu.tpp
     - mc.tpp
     - mc/*.h
     - gpu/*.h
```
The reason of the usage of cholesky_L.h instead of just lower.h is to avoid name conflicts (other functions (e.g. transformation from generalized to std eigenproblem) has upper/lower implementations.).

**Comment by teonnik** [comment]
> Moreover I think that using ETI for matrices and tiles doesn't improve compile time, as only a few members are not inlined.

Currently, that is the case. I checked locally with clang 10 and the overall compile times are very close with or without ETI for `Tile<>` and `Matrix<>`. However, I'd argue that as the number of translation units using `Matrix<>` and `Tile<>` (miniapps and tests) grows, ETI will start having a more noticable effect. One reason it may not be a clear winner yet is because, as we found our recently with `hpx::future<Tile<>>`, ETI of classes instantiates a whole bunch of member functions and constructors that one might not necessarily use. Perhaps this is what is happening with `hpx::promise<>` and `hpx::future<>` as they are stored within the class. This is in contrast to compilers doing implicit instantiations where it appears, only what is needed is instantiated. Perhaps, to alleviate that we can have a more focused approach and ETI only member functions that we suspect contribute significantly to compile times. 

The fact that there are few members of these types that are non-inlined is not a guarantee that these members don't significantly contribute to the compile time... for example, if there was an `hpx::dataflow()` called in any of the member functions, the compile time contribution of that member function will have been rather high. Even for a function like `set_value()` called in the destructor of `~Tile()`, there are several layers of functions belonging to a hierarchy of types that are called. For `Matrix<>` that is even more the case with `operator()` and `read()`, I suspect `hpx::async()` is about as heavy as `hpx::dataflow()`.

> This is not possible all the times, since a template member function of a template class may exist.

You can do ETI on that specific member function if you think it may be worth it.

> The general template should be defined somewhere.

The suggested general template from my comment is as follows:

```
include/dlaf/
    <category>.h
    <category>/
         <operation>.h
         <operation>/
```

where `<operation>/` hosts files internal to the `<operation>`: auxiliary/helper functions and types.

> I'm sure that we need specific GPU implementations.

I was arguing that it is currently not clear how divergent the CPU and GPU implementations are going to be. If the majority of the implementation can be shared, we might end up with a hierarchy of files for one of the backends which is composed of rather small, relatively trivial files, which is something I would like to avoid. The current folder structure brings about a decent amount of complexity and seems to be introduced rather prematurely as we currently only have a CPU backend. I think sticking to a simpler structure until the need for a more complicated one is a better approach overall...  

>
 ```
factorization/
  - factorization.h # includes the backends and declarations of level1 functions.
  - mc.h # declaration of mc backend functions (level2)
  - gpu.h # declaration of gpu backend functions (level2)
  - cholesky/
     - cholesky.tpp # implementation of level1 functions (i.e. the logic that chose which level2 has to be called)
     - mc.tpp # implementation of level2 functions (mc backend) (dispatcher for internal functions)
     - mc/*.h # implementation of internal functions of mc backend
     - gpu/*.h # implementation of internal functions of gpu backend
  - lu/
     - lu.tpp
     - mc.tpp
     - mc/*.h
     - gpu/*.h
```

As discussed previously on the channel, my stance on the whole `.tpp`/`.h` split hasn't changed... 

- I don't think it aids comprehension, quite the reverse, as more files have to be navigated
  - Most of our `.tpp` files are quite short... they won't make our headers a giant mess of code... we rarely even break past 100 lines in them
- It is somewhat unsettling to see an `#include` statement in the middle of your header 
- Breaks most static analysis tools
- You can still have separation of declaration and definition in the header. The point about `.tpp` being an intermediate state to a `.cpp` file is rather moot with ETI as we have it.
- The .h/.cpp split is rooted on whether something is meant to be "inline" or not, templated entities are always "inline", moving a template entity to a `.tpp` file doesn't change that.

If we want to keep separate hierarchies for each backend, then perhaps this alternative can be considered:

```
mc/
  factorization.h
  factorization/
     cholesky.h
     cholesky/  # internals
     lu.h
     lu/  # internals
gpu/
  # same as above
```

> ...<my comment>
>  Currently namespaces are limited to dlaf::::<internal/test/util>. Did I miss overuse somewhere?

> ...<my comment>
> Overload will not be sufficient. It should be possible to use the GPU backend even when the initial data is on CPU memory. Moreover two devices may share the same memory (e.g. APUs)

We should perhaps discuss these in a separate issue. I shouldn't have brought them up here, they are somewhat orthogonal to the topic of this issue.

**Comment by rasolca** [comment]
> > Moreover I think that using ETI for matrices and tiles doesn't improve compile time, as only a few members are not inlined.
> 
> Currently, that is the case. I checked locally with clang 10 and the overall compile times are very close with or without ETI for `Tile<>` and `Matrix<>`. However, I'd argue that as the number of translation units using `Matrix<>` and `Tile<>` (miniapps and tests) grows, ETI will start having a more noticable effect. One reason it may not be a clear winner yet is because, as we found our recently with `hpx::future<Tile<>>`, ETI of classes instantiates a whole bunch of member functions and constructors that one might not necessarily use. Perhaps this is what is happening with `hpx::promise<>` and `hpx::future<>` as they are stored within the class. This is in contrast to compilers doing implicit instantiations where it appears, only what is needed is instantiated. Perhaps, to alleviate that we can have a more focused approach and ETI only member functions that we suspect contribute significantly to compile times.
> 
> The fact that there are few members of these types that are non-inlined is not a guarantee that these members don't significantly contribute to the compile time... for example, if there was an `hpx::dataflow()` called in any of the member functions, the compile time contribution of that member function will have been rather high. Even for a function like `set_value()` called in the destructor of `~Tile()`, there are several layers of functions belonging to a hierarchy of types that are called. For `Matrix<>` that is even more the case with `operator()` and `read()`, I suspect `hpx::async()` is about as heavy as `hpx::dataflow()`.

Good point.

> 
> > This is not possible all the times, since a template member function of a template class may exist.
> 
> You can do ETI on that specific member function if you think it may be worth it.

My question is where to put the implementation when it is not possible (types not known) or it doesn't worth it.

> > 
> 
> ```
> factorization/
>  - factorization.h # includes the backends and declarations of level1 functions.
>  - mc.h # declaration of mc backend functions (level2)
>  - gpu.h # declaration of gpu backend functions (level2)
>  - cholesky/
>     - cholesky.tpp # implementation of level1 functions (i.e. the logic that chose which level2 has to be called)
>     - mc.tpp # implementation of level2 functions (mc backend) (dispatcher for internal functions)
>     - mc/*.h # implementation of internal functions of mc backend
>     - gpu/*.h # implementation of internal functions of gpu backend
>  - lu/
>     - lu.tpp
>     - mc.tpp
>     - mc/*.h
>     - gpu/*.h
> ```
> 
> As discussed previously on the channel, my stance on the whole `.tpp`/`.h` split hasn't changed...
> 
>     * I don't think it aids comprehension, quite the reverse, as more files have to be navigated
>       
>       * Most of our `.tpp` files are quite short... they won't make our headers a giant mess of code... we rarely even break past 100 lines in them
> 
>     * It is somewhat unsettling to see an `#include` statement in the middle of your header
> 
>     * Breaks most static analysis tools
> 
>     * You can still have separation of declaration and definition in the header. The point about `.tpp` being an intermediate state to a `.cpp` file is rather moot with ETI as we have it.
> 
>     * The .h/.cpp split is rooted on whether something is meant to be "inline" or not, templated entities are always "inline", moving a template entity to a `.tpp` file doesn't change that.
 
I'm not meaning that the structure cannot be changed, I just explained why it was chosen to be like this. We can arrange the .tpp problem, but we shouldn't simplify things too much because we have only the CPU backend. We learned a lot from the prototype and we should not ignore it.

> > Currently namespaces are limited to dlaf::::<internal/test/util>. Did I miss overuse somewhere?
Sorry here there is a display problem. I wrote `dlaf::<category>::<internal/test/util>` but it has been interpreted as an html tag.

> > Overload will not be sufficient. It should be possible to use the GPU backend even when the initial data is on CPU memory. Moreover two devices may share the same memory (e.g. APUs)
> 
> We should perhaps discuss these in a separate issue. I shouldn't have brought them up here, they are somewhat orthogonal to the topic of this issue.

Unfortunately the code structure influences the file structure:
To avoid to have `dlaf::category::Function<Backend>::foo(...)`, where foo is usually called `run`, `execute`, ...
we have decided to use `dlaf::Category<Backend>::function(...)`. This requires all functions to be declared where the template specialization of Category is defined. tpp files were useful here since they allow to implement the different functions in different files.

Therefore we should discuss both code and directory structure together.

**Comment by rasolca** [comment]
Please note that #139 is taken in consideration in #154.

**Comment by albestro** [comment]
I don't know if it is the right place where to raise awareness about it, in case let me know and I will fix.

At the moment, I'm facing the problem of adding some building blocks for Factorization, that AFAIK we don't want to expose in the public API. Just to be more clear, let's use this specific case as an example.

I'm working on reduction to band algorithm `dlaf::Eigensolver::reductionToBand<Backend::MC>()` and I have to extract `computeTFactor()` function from it (for the moment it is just one function, but potentially there is at least another one).
Discussing with @rasolca this can be considered a building block of the QR-factorization and it is something we are going to use in other algorithms (e.g. backtransform step of the eigensolver), but we don't want to expose it, as already said.

Since `Factorization<backend>` is a struct, we cannot have a proper `internal` namespace, even if we can workaround it with another nested `struct`. This is the point: how do we want to organize this?

@teonnik @rasolca @msimberg

**Comment by msimberg** [comment]
My 2 cents: It's not clear to me what the purpose of the `Factorization` struct is actually. Why template at that level rather than at the individual algorithm level? From a user perspective I prefer less nesting. It's also not always possible to classify a (sub-)algorithm to be part of only one higher-level operation (or is it always clear in the case of DLAF?). If you do want to keep the `Factorization` struct, what's stopping you from having a `dlaf::internal::Factorization` struct for the internal algorithms?

**Comment by rasolca** [comment]
We need two template levels (as the algorithm is templated on the matrix element type and memory position (Device) and partial specialization on the Backend is not possible). Instead to go for e.g. `SomeStructNameForAllAlgorithms<Backend>::routine<Type, Device>(...)` or e.g. `Cholesky<Backend>::execute<Type>()` we decided to split the algorithms in categories and use them for the backend struct.
Factorization is one of them which includes Cholesky and in the future will include QR, LU, ...
As discussed multiple times before @msimberg joined the team, this is a first attempt to address our needs, but it is far from optimal, and can be change at any time when we find a more suitable structure.
Now that we started to develop the GPU backend it might be easier to find a solution.

**Comment by msimberg** [comment]
I'm not trying to argue one way or the other, so don't take this too seriously. Just putting this out there as another options. It might be useful to make it possible to partially specialize algorithms:

```
namespace internal {
    template <Backend B, Device D, typename Type>
    struct algo_impl;

    // Implementations follow, including partial specializations if needed.
    template <typename Type>
    struct algo_impl<Backend::GPU, Device::GPU, Type> {
        static return_type call(...) {
            // GPU implementation
        }
    }
}

template <Backend B, Device D, typename Type>
auto algo(...) {
    internal::algo_impl<B, D, Type>::call(...);
}
```

This requires slightly more boilerplate, but allows partial specializations.

**Comment by rasolca** [comment]
In fact this looks much better than the current solution... and it can help to simplify the directory structure as well.

@teonnik , @albestro what do you think?

**Comment by albestro** [comment]
This proposal looks very good to me. The additional boilerplate does not look too heavy, it reduces the nesting and, as also @rasolca said, it may ease also the folder structure.

The only additional thing related to this topic may be the explicit template initialization, but I don't see any problem there.

**Comment by teonnik** [comment]
I'm on board with @msimberg 's idea!

With regards to ETI, it is OK to only have it at the lowest level (where the bulk of the implementation is). For example, we do ETI for `cholesky_L()`, but we need not do it for functions calling it (e.g. `Factorization<Backend::MC>::cholesky()` currently).

**Comment by rasolca** [comment]
Closed by #290.
